File: Zelig/Zelig/CompileTime/CodeGenerator/LLVMIR/Module.cpp
#include "llvmheaders.h"
#include "BasicBlock.h"
#include "BasicBlock_impl.h"
#include "Module.h"
#include "Module_impl.h"
#include "Value.h"
#include "Value_impl.h"
#include "Type.h"
#include "Type_impl.h"

#include <io.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <Windows.h>

#include <exception>

NS_MICROSOFT
NS_ZELIG
NS_LLVM

//--//

Module_impl::Module_impl( const char* assembly, unsigned nativeIntSize )
{
    InitializeNativeTarget( );
    _pC = new llvm::LLVMContext( );
    _pM = new Module( assembly, ( *_pC ) );
    _pB = new IRBuilder<>( ( *_pC ) );
    _nativeIntSize = nativeIntSize;

    _pM->addModuleFlag( Module::ModFlagBehavior::Override, "Debug Info Version", DEBUG_METADATA_VERSION );
    dib = new DIBuilder( *_pM );

    // _pM->setDataLayout( "e-S64" );

    dicu = dib->createCompileUnit( dwarf::DW_LANG_lo_user + 0xabc, "out.bc", "", "ZeligIR2LLVMIR", true, "", 0 );
	SetCurrentDIFile("nowhere");
}

llvm::BasicBlock* Module_impl::InsertBlock( llvm::Function* fp, string name )
{
    auto pBB = BasicBlock::Create( _pB->getContext( ), name.c_str( ), fp );
    return pBB;
}

bool Module_impl::DumpToFile( string filename, bool text )
{
    if( GetFileAttributesA( filename.c_str( ) ) != INVALID_FILE_ATTRIBUTES )
    {
        SetFileAttributesA( filename.c_str( ), FILE_ATTRIBUTE_NORMAL );
    }

    int fd_bc = _open( filename.c_str( ), _O_TRUNC | _O_CREAT | _O_BINARY | _O_RDWR );

    if( fd_bc == -1 )
    {
        cout << "failed to create IR file. GetLastError=0x" << hex << GetLastError( ) << endl;
        return false;
    }

    if( text )
    {
        _pM->print( raw_fd_ostream( fd_bc, true ), nullptr );
    }
    else
    {
        WriteBitcodeToFile( _pM, raw_fd_ostream( fd_bc, true ) );
    }
    return true;
}

bool Module_impl::Compile( )
{
    llvm::PassManager PM;

    PM.add( createVerifierPass( ) );
    // Memory To Register
    //PM.add(createPromoteMemoryToRegisterPass());
    //PM.add(createDeadCodeEliminationPass());
    // Make sure everything is still good.
    PM.add( createVerifierPass( ) );

    //PM.add(createPrintModulePass(outs(), "code compilation"));
    bool retVal = PM.run( *_pM );
    dib->finalize( );
    return retVal;
}

void Module_impl::SetCurrentDIFile( std::string fn )
{
    if( dIFiles.find( fn ) == dIFiles.end( ) )
    {
        dIFiles[ fn ] = new DIFile( );
        *dIFiles[ fn ] = dib->createFile( fn, "" );
    }

    curDIFile = dIFiles[ fn ];
}

DISubprogram *Module_impl::GetDISubprogram( std::string fn )
{
    return dISubprograms[ fn ];
}

void Module_impl::SetDISubprogram( std::string fn, DISubprogram *disub )
{
    if( dISubprograms[ fn ] == nullptr )
    {
        dISubprograms[ fn ] = disub;
    }
}

unsigned Module_impl::GetNativeIntSize( )
{
    return _nativeIntSize;
}

unsigned Module_impl::GetPointerSize( )
{
    return GetNativeIntSize( );
}

llvm::Module *Module_impl::GetLLVMObject( )
{
    return _pM;
}

//*************************************************//
//*************************************************//
//*************************************************//

_Module::_Module( String^ assemblyName, unsigned nativeIntSize )
{
    const char* name = (const char*)( Marshal::StringToHGlobalAnsi( assemblyName ) ).ToPointer( );
    _pimpl = new Module_impl( name, nativeIntSize );
    _globalsCounter = 0;
}

void _Module::SetCurrentDIFile( String^ fn )
{
    const char* name = (const char*)( Marshal::StringToHGlobalAnsi( fn ) ).ToPointer( );
    _pimpl->SetCurrentDIFile( name );
}

bool _Module::CheckTypeExistenceByName( String ^name )
{
    const char* strName = (const char*)( Marshal::StringToHGlobalAnsi( name ) ).ToPointer( );
    return _pimpl->GetLLVMObject( )->getTypeByName( strName ) != nullptr;
}

_Type ^_Module::GetOrInsertType( Type_impl *timpl )
{
    return gcnew _Type( this, timpl );
}

_Type ^_Module::GetOrInsertType( String^ name, int sizeInBits )
{
    const char* strName = (const char*)( Marshal::StringToHGlobalAnsi( name ) ).ToPointer( );
    return GetOrInsertType( Type_impl::GetOrInsertTypeImpl( this->_pimpl, strName, sizeInBits ) );
}

_Type ^_Module::GetOrInsertFunctionType( String^ name, _Type ^returnType, List<_Type ^>^ argTypes )
{
    const char* strName = (const char*)( Marshal::StringToHGlobalAnsi( name ) ).ToPointer( );

    std::vector<llvm::Type *> params;

    if( argTypes != nullptr )
    {
        for each( _Type^ type in argTypes )
        {
            params.push_back( type->_pimpl->GetLLVMObjectForStorage( ) );
        }
    }

    llvm::Type *funcType = FunctionType::get( returnType->_pimpl->GetLLVMObjectForStorage( ), params, false );

    Type_impl *timpl = Type_impl::GetOrInsertTypeImpl( this->_pimpl, strName, 0, funcType );

    for each( _Type^ type in argTypes )
    {
        timpl->functionArgs.push_back( type->_pimpl );
    }

    return GetOrInsertType( timpl );
}

_Type ^_Module::GetOrInsertPointerType( String^ name, _Type ^underlyingType )
{
    const char* strName = (const char*)( Marshal::StringToHGlobalAnsi( name ) ).ToPointer( );
    llvm::Type *ptrType;
    //Special case for void pointer
    assert( !underlyingType->_pimpl->GetLLVMObjectForStorage( )->isVoidTy( ) );

    ptrType = underlyingType->_pimpl->GetLLVMObjectForStorage( )->getPointerTo( );

    _Type ^retTy = GetOrInsertType( Type_impl::GetOrInsertTypeImpl( this->_pimpl, strName, _pimpl->GetPointerSize( ), ptrType ) );
    retTy->_pimpl->SetValueTypeFlag( true );
    retTy->_pimpl->SetHasHeaderFlag( false );
    retTy->_pimpl->SetBoxedFlag( false );
    retTy->_pimpl->underlyingPointerType = underlyingType->_pimpl;

    return retTy;
}

_Type ^_Module::GetOrInsertPointerType( _Type ^underlyingType )
{
    return GetOrInsertPointerType( gcnew String( ( underlyingType->_pimpl->GetName( ) + " *" ).c_str( ) ), underlyingType );
}

_Type ^_Module::GetOrInsertBoxedType( String^ name, _Type ^headerType, _Type ^underlyingType )
{
    const char* strName = (const char*)( Marshal::StringToHGlobalAnsi( name ) ).ToPointer( );

    std::vector<llvm::Type*> members;
    members.push_back( headerType->_pimpl->GetLLVMObject( ) );
    members.push_back( underlyingType->_pimpl->GetLLVMObjectForStorage( ) );
    llvm::Type *structType = StructType::create( this->_pimpl->GetLLVMObject( )->getContext( ), members, strName, true );

    _Type ^retTy = GetOrInsertType( Type_impl::GetOrInsertTypeImpl( this->_pimpl, strName, headerType->GetSizeInBits( ) + underlyingType->GetSizeInBitsForStorage( ), structType ) );
    retTy->_pimpl->SetValueTypeFlag( false );
    retTy->_pimpl->SetHasHeaderFlag( true );
    retTy->_pimpl->SetBoxedFlag( true );
    retTy->_pimpl->underlyingBoxedType = underlyingType->_pimpl;

    return retTy;
}

_Type ^_Module::GetOrInsertZeroSizedArray( _Type ^type )
{
    llvm::Type *arrayType = ArrayType::get( type->_pimpl->GetLLVMObjectForStorage( ), 0 );

    _Type ^retTy = GetOrInsertType( Type_impl::GetOrInsertTypeImpl( this->_pimpl, std::string( "MemoryArray of " ) + type->_pimpl->GetName( ), type->GetSizeInBitsForStorage( ), arrayType ) );
    retTy->_pimpl->SetValueTypeFlag( true );
    retTy->_pimpl->SetHasHeaderFlag( false );
    retTy->_pimpl->SetBoxedFlag( false );
    return retTy;
}

_Type ^_Module::GetType( String^ name )
{
    const char* strName = (const char*)( Marshal::StringToHGlobalAnsi( name ) ).ToPointer( );
    Type_impl *timpl = Type_impl::GetTypeImpl( strName );
    if( timpl == nullptr ) return nullptr;
    return GetOrInsertType( timpl );
}

_Type ^_Module::GetVoidType( )
{
    return GetType( "System.Void" );
}

_Value ^_Module::GetIntConstant( _Type ^type, uint64_t v, bool isSigned )
{
    Constant *val = ConstantInt::get( _pimpl->GetLLVMObject( )->getContext( ),
                                      APInt( type->_pimpl->GetSizeInBits( ), v, isSigned ) );

    if( type->_pimpl->GetName( ) == "LLVM.System.IntPtr" || type->_pimpl->GetName( ) == "LLVM.System.UIntPtr" )
    {
        val = ConstantExpr::getIntToPtr( val, type->_pimpl->GetLLVMObject( ) );
    }

    return gcnew _Value( this, new Value_impl( type->_pimpl, val, true ) );
}

_Value ^_Module::GetFloatConstant( float c )
{
    _Type^ t = GetType( "LLVM.System.Single" );
    Value *val = llvm::ConstantFP::get( _pimpl->GetLLVMObject( )->getContext( ), APFloat( c ) );
    return gcnew _Value( this, new Value_impl( t->_pimpl, val, true ) );
}

_Value ^_Module::GetDoubleConstant( double c )
{
    _Type^ t = GetType( "LLVM.System.Double" );
    Value *val = llvm::ConstantFP::get( _pimpl->GetLLVMObject( )->getContext( ), APFloat( c ) );
    return gcnew _Value( this, new Value_impl( t->_pimpl, val, true ) );
}

_Value ^_Module::GetNullPointer( _Type ^type )
{
    Value *val = llvm::Constant::getNullValue( type->_pimpl->GetLLVMObjectForStorage( ) );
    return gcnew _Value( this, new Value_impl( type->_pimpl, val, true ) );
}

bool _Module::CheckIfAExtendsB( _Type ^a, _Type ^b )
{
    llvm::Type *tyA = a->_pimpl->GetLLVMObject( );
    llvm::Type *tyB = b->_pimpl->GetLLVMObject( );

    if( tyA == tyB ) return true;

    if( tyA->isStructTy( ) && tyB->isStructTy( ) )
    {
        StructType *stA = cast<StructType>( tyA );

        if( stA->isOpaque( ) ) return false;
        llvm::Type *stSuper = stA->getStructElementType( 0 );

        while( stSuper->isStructTy( ) )
        {
            if( stSuper == tyB ) return true;
            if( cast<StructType>( stSuper )->isOpaque( ) ) return false;
            stSuper = cast<StructType>( stSuper )->getStructElementType( 0 );
        }
    }
    return false;
}

bool _Module::CheckIfBothTypesPertainToTheSameHierarchy( _Type ^a, _Type ^b )
{
    return CheckIfAExtendsB( a, b ) || CheckIfAExtendsB( b, a );
}

_Function^ _Module::GetOrInsertFunction( String^ name, _Type^ funcType )
{
    return gcnew _Function( this, name, funcType );
}

_UntypedConstValue ^_Module::GetUCVStruct( _Type^ structType, List<_UntypedConstValue ^>^ structMembers, bool anon )
{
    vector<Constant *> fields;

    if( structMembers != nullptr )
    {
        for each( _UntypedConstValue^ ucv in structMembers )
        {
            llvm::Type *curType = cast<StructType>( structType->_pimpl->GetLLVMObject( ) )->getTypeAtIndex( fields.size( ) );
            llvm::Constant *curVal = ucv->val;

            //Zero initializer coersion shortcut:

            if( curVal->isZeroValue( ) )
            {
                curVal = Constant::getNullValue( curType );
            }
            else
            {
                if( curType->isPointerTy( ) && curVal->getType( ) != curType )
                {
                    //It seems that llvm is wrongly switching the bitcast to a GEP.
                    //Let's try to force it with a ptrtoint, inttotr
                    //curVal = ConstantExpr::getPtrToInt( curVal, llvm::IntegerType::get( _pimpl->GetLLVMObject( )->getContext( ), 32 ) );
                    //curVal = ConstantExpr::getIntToPtr( curVal, curType );
                    curVal = ConstantExpr::getBitCast( curVal, curType );
                }
            }

            fields.push_back( curVal );
        }
    }

    llvm::Constant *retVal;
    if( anon )
    {
        retVal = ConstantStruct::getAnon( fields, true );
    }
    else
    {
        retVal = ConstantStruct::get( cast<StructType>( structType->_pimpl->GetLLVMObject( ) ), fields );
    }

    return gcnew _UntypedConstValue( retVal );
}

_UntypedConstValue ^_Module::GetUCVArray( _Type^ arrayMemberType, List<_UntypedConstValue ^>^ arrayMembers )
{
    vector<Constant *> members;
    llvm::Type *curType = arrayMemberType->_pimpl->GetLLVMObjectForStorage( );

    if( arrayMembers != nullptr )
    {
        for each( _UntypedConstValue^ ucv in arrayMembers )
        {
            llvm::Constant *curVal = ucv->val;

            if( curType->isPointerTy( ) && curVal->getType( ) != curType )
            {
                //It seems that llvm is wrongly switching the bitcast to a GEP.
                //Let's try to force it with a ptrtoint, inttotr
                curVal = ConstantExpr::getPtrToInt( curVal, llvm::IntegerType::get( _pimpl->GetLLVMObject( )->getContext( ), 32 ) );
                curVal = ConstantExpr::getIntToPtr( curVal, curType );
                //curVal = ConstantExpr::getBitCast(curVal, curType);
            }

            members.push_back( curVal );
        }
    }

    return gcnew _UntypedConstValue(
        ConstantArray::get( ArrayType::get( curType, members.size( ) ), members ) );
}

_UntypedConstValue ^_Module::GetUCVInt( _Type ^type, uint64_t v, bool isSigned )
{
    return gcnew _UntypedConstValue( Constant::getIntegerValue( type->_pimpl->GetLLVMObject( ),
        APInt( type->_pimpl->GetSizeInBits( ), v, isSigned ) ) );
}

_UntypedConstValue ^_Module::GetUCVNullPointer( _Type ^type )
{
    return gcnew _UntypedConstValue( Constant::getNullValue( type->_pimpl->GetLLVMObjectForStorage( ) ) );
}

_UntypedConstValue ^_Module::GetUCVZeroInitialized( _Type ^type )
{
    return gcnew _UntypedConstValue( Constant::getNullValue( type->_pimpl->GetLLVMObjectForStorage( ) ) );
}

_UntypedConstValue ^_Module::GetUCVConstantPointerFromValue( _Value^ val )
{
    return gcnew _UntypedConstValue( cast<Constant>( val->_pimpl->GetLLVMObject( ) ) );
}

_Value ^_Module::GetGlobalFromUCV( _Type^ type, _UntypedConstValue ^ucv )
{
    char gName[ 256 ];

    sprintf( gName, "G%d", GetMonotonicUniqueId( ) );

    Value *val = new GlobalVariable( *_pimpl->GetLLVMObject( ), ucv->val->getType( ), false, GlobalValue::InternalLinkage, ucv->val, gName );

    return gcnew _Value( this, new Value_impl( type->_pimpl, val, !type->IsValueType( ) ) );
}

_Value ^_Module::GetUninitializedGlobal( _Type^ type )
{
    char gName[ 256 ];

    sprintf( gName, "G%d", GetMonotonicUniqueId( ) );

    Value *val = new GlobalVariable( *_pimpl->GetLLVMObject( ), type->_pimpl->GetLLVMObject( ), false, GlobalValue::InternalLinkage, nullptr, gName );

    return gcnew _Value( this, new Value_impl( type->_pimpl, val, !type->IsValueType( ) ) );
}

void _Module::CreateAlias( _Value^ val, String^ name )
{
    const char* str = (const char*)( Marshal::StringToHGlobalAnsi( name ) ).ToPointer( );
    llvm::Value *llvmVal = val->_pimpl->GetLLVMObject( );
    if( isa<llvm::GlobalValue>( llvmVal ) )
    {
        llvm::GlobalValue *alias = llvm::GlobalAlias::create( str, cast<llvm::GlobalValue>( llvmVal ) );
        alias->setLinkage( GlobalValue::ExternalLinkage );
    }
    else
    {
        printf( "Warning: Ignoring alias \"%s\" because aliasee is not global.", str );
    }
}

bool _Module::Compile( )
{
    return _pimpl->Compile( );
}

bool _Module::DumpToFile( String^ fileName, bool text )
{
    const char* str = (const char*)( Marshal::StringToHGlobalAnsi( fileName ) ).ToPointer( );
    return _pimpl->DumpToFile( str, text );
}

/**********************************/

_UntypedConstValue::_UntypedConstValue( llvm::Constant *val ) :val( val )
{
}

//--//

NS_END
NS_END
NS_END

File: Zelig/Zelig/CompileTime/CodeGenerator/LLVMIR2/_Module.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.Diagnostics;
using Llvm.NET.Values;
using Llvm.NET;
using Llvm.NET.Types;
using Llvm.NET.DebugInfo;
using Microsoft.Zelig.Runtime.TypeSystem;

namespace Microsoft.Zelig.LLVM
{
    public enum OutputFormat
    {
        BitCodeBinary,
        BitCodeSource,
        TargetObjectFile,
        TargetAsmSource
    }

    public class _Module
    {
        public _Module( string assemblyName )
        {
            StaticState.RegisterAll( );

            // REVIEW: Need to deal with how LLVM thinks about threads and context
            // and figure out how to map it to how Zelig system uses threads...
            // for now assuming the LLVM code gen is done on a single thread
            // and ignoring the issue of disposing the Module.
            // TODO: Get target triple from command line instead of hard coding it here.
            var target = Target.FromTriple( "thumbv7m-none-eabi" );
            TargetMachine = target.CreateTargetMachine( "thumbv7m-none-eabi"
                                                      , string.Empty   // CPU
                                                      , string.Empty   // features
                                                      , CodeGenOpt.None // hard code no optimizations for easier debugging for now...
                                                      , Reloc.Static
                                                      , CodeModel.Default
                                                      );

            // REVIEW: What happens when we start interacting with assemblies written in other languages?
            //         At the moment this is setting the entire module to CSharp.
            LlvmModule = new Module( assemblyName, SourceLanguage.CSharp, "out.bc", "ZeligIR2LLVMIR", false, "", 0 )
                        {
                            TargetTriple = TargetMachine.Triple,
                            Layout = TargetMachine.TargetData
                        };

            LlvmModule.AddModuleFlag( ModuleFlagBehavior.Override, Module.DebugVersionValue, Module.DebugMetadataVersion );
            SetCurrentDIFile( "out.bc" );
        }

        public void SetCurrentDIFile( string fn )
        {
            DIFile DIFile;
            if( !DiFiles.TryGetValue( fn, out DIFile ) )
            {
                DIFile = DIBuilder.CreateFile( fn );
                DiFiles.Add( fn, DIFile );
            }

            CurDiFile = DIFile;
        }

        public bool CheckTypeExistenceByName( string name ) => LlvmModule.GetTypeByName( name ) != null;

        public _Type GetOrInsertType( string name, int sizeInBits ) => _Type.GetOrInsertTypeImpl( this, name, sizeInBits );

        public _Type GetOrInsertFunctionType( string name, _Type returnType, List<_Type> argTypes )
        {
            var llvmArgs = argTypes.Select( t => t.GetLLVMObjectForStorage( ) );

            var funcType = LlvmModule.Context.CreateFunctionType( LlvmModule.DIBuilder
                                                                , CurDiFile
                                                                , returnType.GetLLVMObjectForStorage( )
                                                                , llvmArgs
                                                                );
            _Type timpl = _Type.GetOrInsertTypeImpl( this, name, 0, funcType );
            timpl.FunctionArgs.AddRange( argTypes );
            return timpl;
        }

        public _Type GetOrInsertPointerType( string name, _Type underlyingType )
        {
            Debug.Assert( !underlyingType.GetLLVMObjectForStorage( ).IsVoid( ) );

            var ptrType = new DebugPointerType( underlyingType.GetLLVMObjectForStorage( ),  LlvmModule );
            var sizeAndAlign = PointerSize;
            var retVal = _Type.GetOrInsertTypeImpl( this, name, (int)sizeAndAlign, ptrType );
            retVal.IsValueType = true;
            retVal.HasHeader =  false;
            retVal.IsBoxed = false;
            retVal.UnderlyingPointerType = underlyingType;
            return retVal;
        }

        public _Type GetOrInsertPointerType( _Type underlyingType )
        {
            return GetOrInsertPointerType( $"{underlyingType.Name}*", underlyingType );
        }

        public _Type GetOrInsertBoxedType( string name, _Type headerType, _Type underlyingType )
        {
            var structType = new DebugStructType( LlvmModule, name, LlvmModule.DICompileUnit, name );
            var debugFields = new[ ]
                              { new DebugMemberInfo { Type = headerType.DebugType, Index=0, Name="$header", Flags=DebugInfoFlags.Artificial }
                              , new DebugMemberInfo { Type = underlyingType.DebugType, Index=1, Name="m_value", Flags=DebugInfoFlags.Artificial }
                              };

            structType.SetBody( true
                              , LlvmModule
                              , LlvmModule.DICompileUnit
                              , null
                              , 0
                              , DebugInfoFlags.Artificial
                              , debugFields
                              );
            var retVal = _Type.GetOrInsertTypeImpl( this
                                                  , name
                                                  , headerType.SizeInBits + underlyingType.GetSizeInBitsForStorage( )
                                                  , structType
                                                  );
            retVal.IsValueType = false;
            retVal.HasHeader = true;
            retVal.IsBoxed = true;
            retVal.UnderlyingBoxedType = underlyingType;
            return retVal;
        }

        public _Type GetOrInsertZeroSizedArray( _Type type )
        {
            type.GetLLVMObjectForStorage( );
            var arrayType = new DebugArrayType( type.GetLLVMObjectForStorage( ), LlvmModule, 0 );
            var retVal = _Type.GetOrInsertTypeImpl( this, $"MemoryArray of {type.Name}", type.GetSizeInBitsForStorage( ), arrayType );
            retVal.IsValueType = true;
            retVal.HasHeader = false;
            retVal.IsBoxed = false;
            return retVal;
        }

        public _Type GetType( string name ) => _Type.GetTypeImpl( name );

        public _Type GetVoidType( ) => GetType( "System.Void" );

        public _Value GetIntConstant( _Type type, ulong v, bool isSigned )
        {
            Constant val = LlvmModule.Context.CreateConstant( (uint)type.SizeInBits, v, isSigned );
            string name = type.Name;
            if( name == "LLVM.System.IntPtr" || name == "LLVM.System.UIntPtr" )
            {
                val = ConstantExpression.IntToPtrExpression( val, type.DebugType );
            }

            return new _Value( this, type, val, true );
        }

        public _Value GetFloatConstant( float c )
        {
            _Type t = GetType( "LLVM.System.Single" );
            Value val = LlvmModule.Context.CreateConstant( c );
            return new _Value( this, t, val, true );
        }

        public _Value GetDoubleConstant( double c )
        {
            _Type t = GetType( "LLVM.System.Double" );
            Value val = LlvmModule.Context.CreateConstant( c );
            return new _Value( this, t, val, true );
        }

        public _Value GetNullPointer( _Type type )
        {
            var val = type.GetLLVMObjectForStorage( ).GetNullValue( );
            return new _Value( this, type, val, true );
        }

        // REVIEW: Why does code at this low a level or stage need to be concerned with 
        // type hierarchy details? Shouldn't any such issues already be covered in the
        // layers above this?
        public bool CheckIfAExtendsB( _Type a, _Type b )
        {
            ITypeRef tyA = a.DebugType;
            ITypeRef tyB = b.DebugType;

            // REVIEW:
            // While this is from the original C++ it seems a bit odd
            // given the name of this method...
            // A type is generally not considered an extension of itself
            // ( It is possible the original semantic intent of this
            // function is more along the lines of IsAssignmentCompatible(a,b) )
            if( tyA == tyB )
                return true;

            if( tyA.IsStruct() && tyB.IsStruct( ) )
            {
                var stA = ( IStructType )tyA;

                if( stA.IsOpaque )
                    return false;

                ITypeRef tSuper = stA.Members[ 0 ];
                while( tSuper.IsStruct() )
                {
                    if( tSuper == tyB )
                        return true;

                    var stSuper = ( IStructType )tSuper;
                    if( stSuper.IsOpaque )
                        return false;

                    tSuper = stSuper.Members[ 0 ];
                }
            }
            return false;
        }

        public bool CheckIfBothTypesPertainToTheSameHierarchy( _Type a, _Type b )
        {
            return CheckIfAExtendsB( a, b ) || CheckIfAExtendsB( b, a );
        }

        public _Function GetOrInsertFunction( string name, _Type funcType )
        {
            return new _Function( this, name, funcType );
        }

        public Constant GetUCVStruct( _Type structType, List<Constant> structMembers, bool anon )
        {
            List<Constant> fields = new List<Constant>( );
            var llvmStructType = ( IStructType )( structType.DebugType );
            if( structMembers != null )
            {
                foreach( Constant ucv in structMembers )
                {
                    var curVal = ucv;
                    var curType = llvmStructType.Members[ fields.Count ];

                    //Zero initializer coercion shortcut:
                    if( curVal.IsZeroValue )
                    {
                        curVal = curType.GetNullValue( );
                    }
                    else
                    {
                        if( curType.IsPointer( ) && curVal.Type != curType )
                        {
                            curVal = ConstantExpression.BitCast( curVal, curType );
                        }
                    }

                    fields.Add( curVal );
                }
            }

            Constant retVal;
            if( anon )
            {
                retVal = LlvmModule.Context.CreateConstantStruct( true, fields );
            }
            else
            {
                retVal = LlvmModule.Context.CreateNamedConstantStruct( llvmStructType, fields );
            }

            return retVal;
        }

        public Constant GetUCVArray( _Type arrayMemberType, List<Constant> arrayMembers )
        {
            var members = new List<Constant>( );
            ITypeRef curType = arrayMemberType.GetLLVMObjectForStorage( );

            if( arrayMembers != null )
            {
                foreach( Constant ucv in arrayMembers )
                {
                    Constant curVal = ucv;

                    if( curType.IsPointer( ) && curVal.Type != curType )
                    {
                        curVal = ConstantExpression.BitCast( curVal, curType );
                    }

                    members.Add( curVal );
                }
            }

            return ConstantArray.From( curType, members );
        }

        public Constant GetUCVInt( _Type type, ulong v, bool isSigned )
        {
            return LlvmModule.Context.CreateConstant( type.DebugType, v, isSigned );
        }

        //REVIEW: This smells fishy...
        // Both the NullPointer and ZeroInitialized variants are identical (even in the original C++)
        // This seems problematic as, at least in LLVM, there's a distinction between a NullValue
        // (e.g. all zero value of type T) and a NullPointer ( e.g. a value with Type T* and contents all zero)
        public Constant GetUCVNullPointer( _Type type ) => type.GetLLVMObjectForStorage( ).GetNullValue( );

        public Constant GetUCVZeroInitialized( _Type type ) => type.GetLLVMObjectForStorage( ).GetNullValue( );

        public Constant GetUCVConstantPointerFromValue( _Value val ) => ( Constant )val.LlvmValue;

        public _Value GetGlobalFromUCV( _Type type, Constant ucv, bool isConstant )
        {
            var name = $"{type.Name}_{GetMonotonicUniqueId( )}";
            var gv =LlvmModule.AddGlobal( ucv.Type, name );
            gv.IsConstant = isConstant;
            gv.Linkage = Linkage.Internal;
            gv.Initializer = ucv;
            return new _Value( this, type, gv, !type.IsValueType );
        }

        public _Value GetUninitializedGlobal( _Type type )
        {
            var name = $"{type.Name}_{GetMonotonicUniqueId( )}";
            var gv = LlvmModule.AddGlobal( type.DebugType, name );
            gv.IsConstant = false;
            gv.Linkage = Linkage.Internal;
            return new _Value( this, type, gv, !type.IsValueType );
        }

        public void CreateAlias( _Value val, string name )
        {
            Value llvmVal = val.LlvmValue;
            var gv = llvmVal as GlobalObject;
            if( gv != null )
            {
                var alias =LlvmModule.AddAlias( gv, name );
                alias.Linkage = Linkage.External;
            }
            else
            {
                Console.Error.WriteLine( "Warning: Ignoring alias \"{0}\" because aliasee is not global.", name );
            }
        }

        public bool Compile( )
        {
            DIBuilder.Finish( );
            // performing a verify pass here is just a waste of time...
            // Emiting the bit code file will perform a verify
            //string msg;
            //if( !Module.Verify( out msg ) )
            //{
            //    // TODO: we need a better error handling strategy than this!
            //    Console.Error.WriteLine( msg );
            //    throw new ApplicationException( $"LLVM module verification failed:\n{msg}" );
            //}

            // TODO: Generate object file here...
            // REVIEW: What file name should it use for the output?
            return true;
        }

        public bool DumpToFile( string fileName, OutputFormat format )
        {
            DIBuilder.Finish( );
            try
            {
                switch( format )
                {
                case OutputFormat.BitCodeBinary:
                    LlvmModule.WriteToFile( fileName );
                    return true;

                case OutputFormat.BitCodeSource:
                    System.IO.File.WriteAllText( fileName, LlvmModule.AsString( ) );
                    return true;

                case OutputFormat.TargetObjectFile:
                    TargetMachine.EmitToFile( LlvmModule, fileName, CodeGenFileType.ObjectFile );
                    return true;

                case OutputFormat.TargetAsmSource:
                    TargetMachine.EmitToFile( LlvmModule, fileName, CodeGenFileType.AssemblySource );
                    return true;

                default:
                    return false;
                }
            }
            catch( System.IO.IOException ex )
            {
                Console.Error.WriteLine( ex.ToString( ) );
                return false;
            }
        }

        internal uint NativeIntSize => LlvmModule.Layout.IntPtrType( LlvmModule.Context ).IntegerBitWidth;

        internal uint PointerSize => LlvmModule.Layout.IntPtrType( LlvmModule.Context ).IntegerBitWidth;

        internal Context LlvmContext => LlvmModule.Context;

        internal DebugInfoBuilder DIBuilder => LlvmModule.DIBuilder;

        internal DICompileUnit DICompileUnit => LlvmModule.DICompileUnit;

        internal Module LlvmModule { get; }
        
        internal TargetMachine TargetMachine { get; }

        internal DISubProgram GetDISubprogram( string fn )
        {
            DISubProgram retVal;
            if( DiSubPrograms.TryGetValue( fn, out retVal ) )
                return retVal;

            return null;
        }

        internal void SetDISubprogram( string fn, DISubProgram disub )
        {
            if( !DiSubPrograms.ContainsKey( fn ) )
            {
                DiSubPrograms[ fn ] = disub;
            }
        }

        internal DIFile CurDiFile;
        readonly Dictionary<string, DIFile> DiFiles = new Dictionary<string, DIFile>( );
        readonly Dictionary<string, DISubProgram> DiSubPrograms = new Dictionary<string, DISubProgram>( );

        static int GetMonotonicUniqueId( )
        {
            return ( int )System.Threading.Interlocked.Increment( ref globalsCounter );
        }

        static Int64 globalsCounter;
    }
}
